feat: notify admin to update solvers in the community

Anyitechs opened the pull request:
This PR schedules a daily job at midnight to notify the admin of a community that there are no solvers in the community. The maximum number of notification messages right now is set to 5, afterwards the community is disabled.

Fixes #314 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a more frequent automated check that monitors community setups and notifies administrators when communities lack required solvers.
	- Added a configuration for managing the maximum number of notifications sent to administrators.
	- Implemented multilingual alerts to provide clear guidance to users regarding community management and solver requirements.

- **Bug Fixes**
	- Enhanced error handling for community notifications to ensure proper logging and user guidance.

- **Documentation**
	- Updated locale files to include new messages for users about community solver requirements in multiple languages.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Catrya reviewed (changes requested):
Hello, did you tested it locally? It's not working properly:
- This message is not being displayed: `Your community ${community.name} doesn't have any solvers. Please add at least one solver to avoid being disabled.`
- Even though the logs say: `Community: ${community.name} has been disabled due to lack of solvers.` the community is not being disabled. You can continue using it normally.
- I recommend putting the warning messages to the administrators in the language they have the bot set in, in the yaml of ./locales. You can see as an example at #563

Anyitechs commented:
> Hello, did you tested it locally? It's not working properly:
> 
> * This message is not being displayed: `Your community ${community.name} doesn't have any solvers. Please add at least one solver to avoid being disabled.`
> * Even though the logs say: `Community: ${community.name} has been disabled due to lack of solvers.` the community is not being disabled. You can continue using it normally.
> * I recommend putting the warning messages to the administrators in the language they have the bot set in, in the yaml of ./locales. You can see as an example at [Add cronjob to notify admin about missing solvers (#314) #563](https://github.com/lnp2pBot/bot/pull/563)

Thank you for the review. I'll work on these feedbacks.

Anyitechs commented:
Hi @Catrya, I've addressed all the feedbacks here. Let me know if there're other changes.

> Hello, did you tested it locally? It's not working properly:
> 
> * This message is not being displayed: `Your community ${community.name} doesn't have any solvers. Please add at least one solver to avoid being disabled.`
> * Even though the logs say: `Community: ${community.name} has been disabled due to lack of solvers.` the community is not being disabled. You can continue using it normally.
> * I recommend putting the warning messages to the administrators in the language they have the bot set in, in the yaml of ./locales. You can see as an example at [Add cronjob to notify admin about missing solvers (#314) #563](https://github.com/lnp2pBot/bot/pull/563)

Catrya reviewed (commented):
Hi @Anyitechs , much better, it already sends messages to the admins and disables the community. However, I think the message for the admins could be improved so that they know how much time they have to put a new admin.
I think it could be something like this:

Your community ${communityName} does not have any solvers. Please add at least one within X days to prevent the community from being disabled.

Actually the number of notifications that the solver receives before disabling the community is MAX_MESSAGES -1 because it disables it in MAX_MESSAGES, I suggest that it shows that number of notifications  before disabling it

I will continue testing and I will tell you if I find any bug, good job!

Anyitechs commented:
> Hi @Anyitechs , much better, it already sends messages to the admins and disables the community. However, I think the message for the admins could be improved so that they know how much time they have to put a new admin. I think it could be something like this:
> 
> Your community ${communityName} does not have any solvers. Please add at least one within X days to prevent the community from being disabled.
> 
> Actually the number of notifications that the solver receives before disabling the community is MAX_MESSAGES -1 because it disables it in MAX_MESSAGES, I suggest that it shows that number of notifications before disabling it
> 
> I will continue testing and I will tell you if I find any bug, good job!

Thank you for the feedback @Catrya. I just pushed a new commit with the changes requested here. Please take a look and let me know your thoughts. Thank you.

Catrya reviewed (approved):
ACK. Thanks @Anyitechs !

Catrya reviewed (commented):
Hi @Anyitechs, sorry but I remembered that when a community is disabled it should also be deleted from the database, see `deleteCommunity`
This PR does not delete the community from the database, and I think that is necessary.
Anyway, let's see what @grunch thinks, sorry

Anyitechs commented:
> Hi @Anyitechs, sorry but I remembered that when a community is disabled it should also be deleted from the database, see `deleteCommunity` This PR does not delete the community from the database, and I think that is necessary. Anyway, let's see what @grunch thinks, sorry

Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first? 

Any thoughts? @grunch

Catrya commented:
> Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first?
> 
> Any thoughts? @grunch

Hi @Anyitechs , I don't mean to disable it first and then delete it, but to delete it.
Send messages to users as you already do, but if admin doesn't add a solver, just delete it from the database (we already called that disabling)

Anyitechs commented:
> > Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first?
> > Any thoughts? @grunch
> 
> Hi @Anyitechs , I don't mean to disable it first and then delete it, but to delete it. Send messages to users as you already do, but if admin doesn't add a solver, just delete it from the database (we already called that disabling)

That makes sense. Thank you for the clarification. I'll work on that and update the branch.

Anyitechs commented:
@Catrya I just pushed an update to the branch now. Please do take a look and let me know your thoughts. Thank you.

Catrya reviewed (approved):
thanks @Anyitechs I already tested it and it works well to me! Let's see @grunch review

Anyitechs commented:
> thanks @Anyitechs I already tested it and it works well to me! Let's see @grunch review

Thank you @Catrya. @grunch let me know if there're any changes.

grunch commented on bot/modules/community/commands.js:
```diff
@@ -143,7 +143,7 @@ exports.updateCommunity = async (ctx, id, field, bot) => {
     if (!(await validateObjectId(ctx, id))) return;
     const community = await Community.findOne({
       _id: id,
-      creator_id: user._id,
```
as you can read in our contribution section, we are using the airbnb javascript guide style, please don't remove the comma

https://github.com/airbnb/javascript?tab=readme-ov-file#commas--dangling

Anyitechs replied:
Adding them back now.

grunch commented on bot/modules/community/commands.js:
```diff
@@ -211,7 +211,7 @@ exports.deleteCommunity = async ctx => {
     const community = await Community.findOne({
       _id: id,
-      creator_id: ctx.user._id,
+      creator_id: ctx.user._id
```
https://github.com/airbnb/javascript?tab=readme-ov-file#commas--dangling

grunch commented on bot/modules/community/commands.js:
```diff
@@ -233,7 +233,7 @@ exports.changeVisibility = async ctx => {
     const community = await Community.findOne({
       _id: id,
-      creator_id: ctx.user._id,
+      creator_id: ctx.user._id
```
[as you can read in our contribution section, we are using the airbnb javascript guide style, please don't remove the comma

](https://github.com/airbnb/javascript?tab=readme-ov-file#commas--dangling)

grunch commented on jobs/check_solvers.ts:
```diff
@@ -0,0 +1,56 @@
+
+    await bot.telegram.sendMessage(
+      admin.tg_id,
+      message
```
https://github.com/airbnb/javascript?tab=readme-ov-file#commas--dangling

knocte commented:
> as you can read in our contribution section, we are using the airbnb javascript guide style, please don't remove the comma

How about moving to a style that can be checked at CI time? e.g. with prettier

knocte commented:
@webwarrior-ws hey, please check CI, it is red.

knocte commented:
> @webwarrior-ws hey, please check CI, it is red.

Oops wrong contributor! Nevermind.

grunch commented:
> > as you can read in our contribution section, we are using the airbnb javascript guide style, please don't remove the comma
> 
> How about moving to a style that can be checked at CI time? e.g. with prettier

yes I have it locally but we need it at CI level

Anyitechs commented:
Thank you for the review @grunch. I just addressed the feedbacks, rebased and updated the branch.

grunch commented:
> > Hi @Anyitechs, sorry but I remembered that when a community is disabled it should also be deleted from the database, see `deleteCommunity` This PR does not delete the community from the database, and I think that is necessary. Anyway, let's see what @grunch thinks, sorry
> 
> Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first?
> 
> Any thoughts? @grunch

I agree on this, we allow users to disable a community for a while, maybe they need to stop it and want to go back later, but we already have a cronjob that deletes a community after N days without a successful order, disabled or not, so I think there is no harm if someone disable a community and forget about it, it will be deleted at some point

Catrya commented:
> > > Hi @Anyitechs, sorry but I remembered that when a community is disabled it should also be deleted from the database, see `deleteCommunity` This PR does not delete the community from the database, and I think that is necessary. Anyway, let's see what @grunch thinks, sorry
> > 
> > 
> > Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first?
> > Any thoughts? @grunch
> 
> I agree on this, we allow users to disable a community for a while, maybe they need to stop it and want to go back later, but we already have a cronjob that deletes a community after N days without a successful order, disabled or not, so I think there is no harm if someone disable a community and forget about it, it will be deleted at some point

Anyway, @Anyitechs then got the community deleted, which is better IMHO
I'm going to retest everything asap

grunch commented:
> > > > Hi @Anyitechs, sorry but I remembered that when a community is disabled it should also be deleted from the database, see `deleteCommunity` This PR does not delete the community from the database, and I think that is necessary. Anyway, let's see what @grunch thinks, sorry
> > > 
> > > 
> > > Thank you @Catrya. I was thinking we needed a soft delete and that was why I added the `is_disabled` flag to the community model. But why do we need to disable a community first and also delete it from the database. Why don't we go ahead and delete the community rather than having to disable it first?
> > > Any thoughts? @grunch
> > 
> > I agree on this, we allow users to disable a community for a while, maybe they need to stop it and want to go back later, but we already have a cronjob that deletes a community after N days without a successful order, disabled or not, so I think there is no harm if someone disable a community and forget about it, it will be deleted at some point
> 
> Anyway, @Anyitechs then got the community deleted, which is better IMHO
> I'm going to retest everything asap

I'm sorry I was thinking you were talking about adding a new way of just disable a community from the same user creator 

I agree on, if the creator execute the command to delete the community it should delete the community inmediatly

Catrya commented on .env-sample:
```diff
@@ -80,4 +80,7 @@ COMMUNITY_CURRENCIES=20
\ No newline at end of file
+DISPUTE_START_WINDOW=600
+
+# Number of notification messages sent to the admin, informing them of lack of solvers before disabling the community
```
The admin is being sent `MAX_MESSAGES -1` notifications, if it says 10 here, the admin will see 9 notifications.
I think that either this description, or `MAX_MESSAGES `  needs to be adjusted to make it clearer for whoever is going to configuring it. 

For the rest, everything seems fine to me.

Anyitechs replied:
Thank you @Catrya. What do you recommend? 

Adjusting the description to include that the admin receives `MAX_MESSAGES -1` notifications seems ok to me.

Catrya replied:
Yes, that's fine by me too, go ahead. Thanks @Anyitechs

Anyitechs replied:
Just updated the description and rebased.

Catrya commented:
Hi @Anyitechs Have you tested it? It doesn't work for me with the latest commit

Anyitechs commented:
> Hi @Anyitechs Have you tested it? It doesn't work for me with the latest commit

Hi @Catrya, what issue did you notice? I tested and it works. The latest commit is on documentation, but I tested again and it works. Let me know what issue you noticed so I can try to reproduce it here.

Catrya reviewed (approved):
tACK

Catrya commented:
> Hi @Catrya, what issue did you notice? I tested and it works. The latest commit is on documentation, but I tested again and it works. Let me know what issue you noticed so I can try to reproduce it here.

Hi @Anyitechs  sorry, that was a mistake on my part. I tested again and I already approved it, it works fine. Good job. Congratulations

Catrya commented:
> Hi @Catrya, what issue did you notice? I tested and it works. The latest commit is on documentation, but I tested again and it works. Let me know what issue you noticed so I can try to reproduce it here.

Hi @Anyitechs  sorry, that was a mistake on my part. I tested again and I already approved it, it works fine. Good job. Congratulations

Catrya commented:
@Anyitechs  before merging, it is necessary that no test fails, could you please check that 1 test is failing and fix it? Thanks
`should schedule recurring jobs`

Anyitechs commented:
> @Anyitechs before merging, it is necessary that no test fails, could you please check that 1 test is failing and fix it? Thanks `should schedule recurring jobs`

Will check it out and fix it.

Catrya reviewed (approved):
tACK thsnks @Anyitechs. Good job!
just wait for @grunch

knocte commented:
Hey! I have couple of suggestions:

knocte commented on .env-sample:
```diff
@@ -80,4 +80,7 @@ COMMUNITY_CURRENCIES=20
+DISPUTE_START_WINDOW=600
+
+# Number of notification messages sent to the admin, informing them of lack of solvers before disabling the community. The admin receives `MAX_MESSAGES - 1` notification messages.
+MAX_MESSAGES=10
```
@Anyitechs I would recommend renaming this to have a much clearer name; right now "messages" sounds too generic, how about MAX_WARNINGS_TO_ADMIN_BEFORE_COMMUNITY_DEACTIVATION

knocte commented on models/community.ts:
```diff
@@ -81,6 +82,7 @@ const CommunitySchema = new Schema<ICommunity>({
     },
     created_at: { type: Date, default: Date.now },
     nostr_public_key: { type: String },
+    messages_sent_count: { type: Number, default: 0 },
```
@Anyitechs I'd suggest renaming this to `warning_messages_count`

Anyitechs commented:
> Hey! I have couple of suggestions:

Thank you for the suggestions @knocte. I've effected them. Let me know if everything looks good.

CC: @Catrya @grunch

Catrya commented on bot/start.ts:
```diff
@@ -200,7 +200,10 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
   });
 
-  schedule.scheduleJob(`0 0 * * *`, async () => {
+  // schedule.scheduleJob(`0 0 * * *`, async () => {
+  //   await checkSolvers(bot);
+  // });
+  schedule.scheduleJob(`*/1 * * * *`, async () => {
```
Hi @Anyitechs Why to change it to check every minute? It should stay at once a day.

Catrya commented on bot/start.ts:
```diff
@@ -200,7 +200,10 @@ const initialize = (botToken: string, options: Partial<Telegraf.Options<MainCont
   });
 
-  schedule.scheduleJob(`0 0 * * *`, async () => {
+  // schedule.scheduleJob(`0 0 * * *`, async () => {
+  //   await checkSolvers(bot);
+  // });
+  schedule.scheduleJob(`*/1 * * * *`, async () => {
```
Hi @Anyitechs Why to change it to check every minute? It should stay at once a day.

Anyitechs replied:
Yes, I've addressed it. Thank you.

Catrya reviewed (approved):
Good job @Anyitechs 
tACK @grunch

grunch reviewed (approved):
LGTM

grunch commented:
Great job @Anyitechs please sign your commits on next contributions 😃

Anyitechs commented:
> Great job @Anyitechs please sign your commits on next contributions 😃

Sure! Thank you @grunch.

grunch commented:
@Anyitechs  This issue was in the rewards board, Please send us a lightning invoice with amount of 400000 sats as a comment in this PR

Anyitechs commented:
> @Anyitechs This issue was in the rewards board, Please send us a lightning invoice with amount of 400000 sats as a comment in this PR

Here @grunch, thank you so much.

lnbc4m1p5q3utfpp5k6lylh440vatgm37sqkqvpn8xxxv8xzeqezl8skwxh462xnl0r3sdqgf3h8qvnscqzzsxqyz5vqsp5qqh8hyvkjuc7tpgaw0cnvpynmh97naezeamswg8wq7jwy70040ds9qxpqysgqud80ycdpqzr5d87nmr39ananj7apfmrsryzjxy9g3j7d384s2jasnw3p0haj4gtqcuk5hh5qe5ezqkpuhua5upcyyttgxvgdwa7v9scpnyc4lx

grunch commented:
> lnbc4m1p5q3utfpp5k6lylh440vatgm37sqkqvpn8xxxv8xzeqezl8skwxh462xnl0r3sdqgf3h8qvnscqzzsxqyz5vqsp5qqh8hyvkjuc7tpgaw0cnvpynmh97naezeamswg8wq7jwy70040ds9qxpqysgqud80ycdpqzr5d87nmr39ananj7apfmrsryzjxy9g3j7d384s2jasnw3p0haj4gtqcuk5hh5qe5ezqkpuhua5upcyyttgxvgdwa7v9scpnyc4lx

Payment hash:   b6be4fdeb57b3ab46e3e802c060667318cc398590645f3c2ce35eba51a7f78e3
Payment status: SUCCEEDED, preimage: bcfe73b06da11947d9f37bb30c51c2f86a9b1e82a64e2184c1f54c53146fa489
